### ISSUE TITLE
 error LNK2019: 无法解析的外部符号 "private: static class RegistrationToken::RegistrationToken const mediapipe::tool::OptionsRegistry::registration_token<class mediapipe::FaceDetectionOptions>"

### ISSUE THREAD

**Issue by jinfagang** [google/mediapipe#4060]
error LNK2019: 无法解析的外部符号 "private: static class RegistrationToken::RegistrationToken const mediapipe::tool::OptionsRegistry::registration_token<class mediapipe::FaceDetectionOptions>"

**Comment by jaydeepsaravaiya** [comment]
This is due to probably a static variable not initialized in a c++ class (RegistrationToken::RegistrationToken const mediapipe::tool::OptionsRegistry).
need to investigate further.
> inference_calculator_options_registry.lo.lib(inference_calculator_options_lib.obj) : error LNK2019: unresolved external symbol "private: static class RegistrationToken::RegistrationToken const mediapipe::tool::OptionsRegistry::registration_token\<class mediapipe::InferenceCalculatorOptions\>" (??$registration_token@VInferenceCalculatorOptions@mediapipe@@@OptionsRegistry@tool@mediapipe@@0VRegistrationToken@3@B) 
> referenced in function "??__E?$registration_token@VInferenceCalculatorOptions@mediapipe@@@OptionsRegistry@tool@mediapipe@@YAXXZ" (??__E?$registration_token@VInferenceCalculatorOptions@mediapipe@@@OptionsRegistry@tool@mediapipe@@YAXXZ)
> bazel-out\x64_windows-opt\bin\mediapipe\examples\desktop\hand_tracking\hand_tracking_cpu.exe : fatal error LNK1120: 1 unresolved externals

**Comment by diegomacario** [comment]
I'm seeing the exact same problem as @jinfagang when I try to build the MediaPipe Hands example on Windows. I was able to build the Hello World example successfully, but not this one.

**Comment by diegomacario** [comment]
I worked around this problem by commenting these two lines in `mediapipe\framework\tool\options_registry.h`:

```
template <class MessageT>
static const RegistrationToken registration_token;
```

And by commenting this in `mediapipe\framework\tool\options_lib_template.cc`:

```
namespace mediapipe {
// The protobuf descriptor for an options message type.
template <>
const RegistrationToken tool::OptionsRegistry::registration_token<
    MP_OPTION_TYPE_NS::MP_OPTION_TYPE_NAME> =
    tool::OptionsRegistry::Register(ReadFileDescriptorSet(
        std::string(kDescriptorContents, sizeof(kDescriptorContents) - 1)));
}  // namespace mediapipe
```

Once I did that, the MediaPipe Hands example built successfully and worked perfectly.

This is a hacky workaround, but I'll share it here in case it helps anyone.

**Comment by jinfagang** [comment]
@diegomacario  waw, nice catch, I will test on my side. I am building face detection on desktop.

**Comment by jinfagang** [comment]
@diegomacario @jaydeepsaravaiya  I think it because of the VirtualStudio version problem. for 2022 version, it fails. 2019 can build.

**Comment by diegomacario** [comment]
Oh good to know, thanks @jinfagang!

**Comment by jinfagang** [comment]
I can verify it was MSVC 2022 failed, @google-admin  Please fix mediapipe MSVC 2022 build, since we are in 2023 now.

**Comment by jaydeepsaravaiya** [comment]
> I'm seeing the exact same problem as @jinfagang when I try to build the MediaPipe Hands example on Windows. I was able to build the Hello World example successfully, but not this one.

Did Hello world worked properly for you. For me it builds but doesnt do anything.

**Comment by ayushgdev** [comment]
Hello @jinfagang Is your issue resolved?

**Comment by jinfagang** [comment]
Yes. but another issue happend.

**Comment by rajkundu** [comment]
This seems to be an issue with VS 2022 17.4+. I've filed a bug report with some more details [here](https://developercommunity.visualstudio.com/t/Static-Data-Member-Templates:-DLL-Linkin/10288174). If you can, please consider upvoting it to ensure that we can build MediaPipe with future versions of VS!

**Comment by rajkundu** [comment]
I can confirm that the underlying bug in VS 2022 and has been fixed in version 17.6 Preview 1. So, the permanent solution is to avoid using VS 2022 17.4.x & 17.5.x and either use VS 2022 <=17.3 or >=17.6 upon release.
